Tasks
Hypothesis based EDA:

Which brand (or make) has more share of vehicles with less than 8.0 fuel consumption?

Plot the correlation scatter plot for the engine size and fuel consumption.

Create a pairplot with all the numerical variables and note down the observations/insights

Create a new column named “Transmission type” from the TRANSMISSION column by separating out the gears from the alphabets. (Hint: use regex)

Build Linear regression model with only the numerical features using Statsmodels library. Note down the R2, Adjusted R2. 

One-hot encode categorical variables and Build Linear regression model using Statsmodels library. Note down the R2, Adjusted R2. Compare it with the previous model.


In [1]:
#import packages
from sklearn import linear_model
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
import statsmodels.api as sm
import seaborn as sns
from scipy import stats
import scipy

In [2]:
# Load the data as a data frame by using local path
data_path="/Users/guilhembarroyer/Desktop/Projects/Machine Learning Path/Project2/Fuel_consumption_prediction/Fuel_Consumption_2000-2022.csv"
df=pd.read_csv(data_path)

In [3]:
#check the columns 
df.columns

Index(['YEAR', 'MAKE', 'MODEL', 'VEHICLE CLASS', 'ENGINE SIZE', 'CYLINDERS',
       'TRANSMISSION', 'FUEL', 'FUEL CONSUMPTION', 'HWY (L/100 km)',
       'COMB (L/100 km)', 'COMB (mpg)', 'EMISSIONS'],
      dtype='object')

In [7]:
df.head()

YEAR   MAKE    MODEL VEHICLE CLASS  ENGINE SIZE  CYLINDERS TRANSMISSION  \
0  2000  ACURA    1.6EL       COMPACT          1.6          4           A4   
1  2000  ACURA    1.6EL       COMPACT          1.6          4           M5   
2  2000  ACURA    3.2TL      MID-SIZE          3.2          6          AS5   
3  2000  ACURA    3.5RL      MID-SIZE          3.5          6           A4   
4  2000  ACURA  INTEGRA    SUBCOMPACT          1.8          4           A4   

  FUEL  FUEL CONSUMPTION  HWY (L/100 km)  COMB (L/100 km)  COMB (mpg)  \
0    X               9.2             6.7              8.1          35   
1    X               8.5             6.5              7.6          37   
2    Z              12.2             7.4             10.0          28   
3    Z              13.4             9.2             11.5          25   
4    X              10.0             7.0              8.6          33   

   EMISSIONS  
0        186  
1        175  
2        230  
3        264  
4        198

In [8]:
df.describe()

YEAR   ENGINE SIZE     CYLINDERS  FUEL CONSUMPTION  \
count  22556.000000  22556.000000  22556.000000      22556.000000   
mean    2011.554442      3.356646      5.854141         12.763513   
std        6.298269      1.335425      1.819597          3.500999   
min     2000.000000      0.800000      2.000000          3.500000   
25%     2006.000000      2.300000      4.000000         10.400000   
50%     2012.000000      3.000000      6.000000         12.300000   
75%     2017.000000      4.200000      8.000000         14.725000   
max     2022.000000      8.400000     16.000000         30.600000   

       HWY (L/100 km)  COMB (L/100 km)    COMB (mpg)     EMISSIONS  
count    22556.000000     22556.000000  22556.000000  22556.000000  
mean         8.919126        11.034341     27.374534    250.068452  
std          2.274764         2.910920      7.376982     59.355276  
min          3.200000         3.600000     11.000000     83.000000  
25%          7.300000         9.100000     22.000000    209.000000  
50%          8.400000        10.600000     27.000000    243.000000  
75%         10.200000        12.700000     31.000000    288.000000  
max         20.900000        26.100000     78.000000    608.000000

In [19]:
average_consumption_fuel=df.groupby('MAKE')['FUEL CONSUMPTION'].mean()
low_consumption_brands=average_consumption_fuel[average_consumption_fuel<=8.0]

print(low_consumption_brands)

MAKE
SMART    5.972727
Name: FUEL CONSUMPTION, dtype: float64


In [21]:
# Créer une colonne booléenne (True si consommation <= 0.8, False sinon)
df['reduced_consumption'] = df['FUEL CONSUMPTION'] <= 8.0

df.describe()
# Regrouper par marque et calculer la proportion de véhicules avec consommation réduite
grouped = df.groupby('MAKE')['reduced_consumption'].mean()


# Filtrer les marques avec plus de 50% de véhicules à consommation réduite
brands_with_low_consumption = grouped[grouped > 0.5]

# Afficher les résultats
print(brands_with_low_consumption)

MAKE
SMART    1.0
Name: reduced_consumption, dtype: float64


SMART is the only brand with more than the half of its cars which consume less than 8.0 fuel. The mean of the fuel consumption of its cars is 5.972727.